src/libsync/owncloudpropagator.h

```diff
@@ -187,27 +221,41 @@ class OWNCLOUDSYNC_EXPORT PropagateDirectory : public PropagatorJob {
     // e.g: create the directory
     QScopedPointer<PropagateItemJob>_firstJob;
 
-    // all the sub files or sub directories.
-    QVector<PropagatorJob *> _subJobs;
+    // all the sub files or sub directories. This map has to be updated with _containerJobs
+    // QMap is ordered by the key value, or in case of equal keys (e.g 0) by insertion order.
+    QMap<quint64, PropagatorJob *> _subJobs;
+
+    QVector<PropagatorJob *> _containerJobs;
 
     SyncFileItemPtr _item;
 
     int _jobsFinished; // number of jobs that have completed
+    int _totalJobs; // number of jobs that will be defined to be synced
     int _runningNow; // number of subJobs running right now
     SyncFileItem::Status _hasError;  // NoStatus,  or NormalError / SoftError if there was an error
-    int _firstUnfinishedSubJob;
+
+    /*
+     * Keeps track of the all the sub files or sub directories priority.
+    */
+    qint64 _subJobsPriority;
 
     explicit PropagateDirectory(OwncloudPropagator *propagator, const SyncFileItemPtr &item = SyncFileItemPtr(new SyncFileItem))
-        : PropagatorJob(propagator)
-        , _firstJob(0), _item(item),  _jobsFinished(0), _runningNow(0), _hasError(SyncFileItem::NoStatus), _firstUnfinishedSubJob(0)
+        : PropagatorJob(propagator, JobPriority::ContainerItemsPriority)
+        , _firstJob(0), _item(item),  _jobsFinished(0),  _totalJobs(0), _runningNow(0), _hasError(SyncFileItem::NoStatus), _subJobsPriority(0)
     { }
 
     virtual ~PropagateDirectory() {
         qDeleteAll(_subJobs);
+        qDeleteAll(_containerJobs);
     }
 
     void append(PropagatorJob *subJob) {
-        _subJobs.append(subJob);
+        _subJobsPriority += subJob->getJobPredicateValue();
```

Comment:
updateJobPredicateValue is recursively traversing only _containerJobs tree, thus only directories with changes. This one will be called once for the _root job, and recursively visit all "changed" directories. _subJobs are untouched, and I will not touch that, because this could contain thousands of items, while _containerJobs will never be deep.